### ISSUE TITLE
Removing tile from a tilemap doesn't unset the tile in the tilemap

### ISSUE THREAD

**Issue by nightblade9** [godotengine/godot#51303]
### Godot version

3.3.2

### System information

Windows 10, GLES3

### Issue description

After removing a tile from a tilemap, all tiles that used to be that tile ID, are still that tile ID - even though the editor shows them as empty.

### Steps to reproduce

- Create a tilemap
- Create a tileset with two tiles (e.g. A and B) and assign it to the tilemap
- Draw some tiles of both A and B on the tilemap
- Edit the tilemap and delete tile B
- Using GDscript, print out the tile_id of every tile

Expected: you'll see tiles with ID 0 only (tile A)
Actual: you see tiles with ID 0 and ID 1 (formerly tile B's ID)

**Comment by kleonc** [comment]
>Removing tile from a tilemap doesn't unset the tile in the tilemap

If you meant _"Removing tile from a TileSet doesn't unset the tile in the TileMap"_ then I'd say that's expected/correct behavior.
If you want to clear  such tiles you can do it by calling [`fix_invalid_tiles()`](https://docs.godotengine.org/en/3.4/classes/class_tilemap.html#class-tilemap-method-fix-invalid-tiles) method on given `TileMap`. You can also trigger it from the editor:
![godot windows tools 64_FnJfKupEL1](https://user-images.githubusercontent.com/9283098/128548135-ea83511e-08f3-4b93-b89a-6a649088ee4c.png)

**Comment by nightblade9** [comment]
> then I'd say that's expected/correct behavior.

I don't think this is expected nor correct. Here's what I did:

- Make a tilemap with three tiles, and draw a bunch of all three. One of them is a "walkable" tile.
- Decide that I like parallax backgrounds instead
- Delete the "walkable" tile from the tileset
- Look at the tilemap. Those tiles previously containing the walkable tile, look empty in the Godot editor.

When I ran my game, it took me a while to figure out why monsters aren't moving onto those tiles - because they're not walkable (tile ID of -1). They have a tile ID of two, which isn't even a valid tile in my tileset.

The work-around was to flood-fill them and flood-fill-erase them (right-click).

**Comment by nightblade9** [comment]
I didn't see until now you mentioned the editor option to fix invalid tiles. That's great. That also relies on you thinking/realizing that there are invisible tiles, which is made more difficult because the editor shows you empty space after deleting an invalid tile. There isn't even a warning/error printed to tip you off that there's something wrong.

**Comment by akien-mga** [comment]
AFAIK the new tiles editor in 4.0 will show you "removed" tiles visually as invalid: https://godotengine.org/article/tiles-editor-progress-report-2

So I think this can be considered fixed in 4.0.

If someone wants to work on a similar UX improvement for 3.x, that could likely be done (draw "removed" tiles with something that makes it clear that they're different from an empty tile (ID `-1`).

Otherwise yes, it's intended that when tiles get removed from a tileset, it doesn't automatically unset them from all tilemaps that use this tileset. This lets you *replace* a tile by a new one easily, or even swap tilesets, without losing information.

**Comment by kleonc** [comment]
> That also relies on you thinking/realizing that there are invisible tiles, which is made more difficult because the editor shows you empty space after deleting an invalid tile. There isn't even a warning/error printed to tip you off that there's something wrong.

Yeah, no feedback to the user is the problem here, not clearing invalid tiles automatically isn't.

---

> AFAIK the new tiles editor in 4.0 will show you "removed" tiles visually as invalid: https://godotengine.org/article/tiles-editor-progress-report-2

Also: [a tweet about it](https://twitter.com/gr0ud/status/1336352585452834821).

> If someone wants to work on a similar UX improvement for 3.x, that could likely be done (draw "removed" tiles with something that makes it clear that they're different from an empty tile (ID -1).

I'll try doing something similar.
